src/macros.rs

```diff
@@ -507,120 +289,65 @@ macro_rules! arg_enum {
-                match *self {
-                    $($e::$v => write!(f, stringify!($v)),)+
-                }
-            }
-        }
-
-        impl $e {
-            #[allow(dead_code)]
-            pub fn variants() -> Vec<&'static str> {
-                vec![
-                    $(stringify!($v),)+
-                ]
-            }
-        }
+    (pub enum $e:ident { $($v:ident),+ } ) => {
+        arg_enum!(@impls
+            (pub enum $e {
+                $($v),+
+            }) -> ($e, $($v),+)
+        );
     };
-    (#[derive($($d:ident),+)] pub enum $e:ident { $($v:ident),+ } ) => {
-        #[derive($($d,)+)]
-        pub enum $e {
-            $($v),+
-        }
-
-        impl ::std::str::FromStr for $e {
-            type Err = String;
-
-            fn from_str(s: &str) -> Result<Self,Self::Err> {
-                use ::std::ascii::AsciiExt;
-                match s {
-                    $(stringify!($v) |
-                    _ if s.eq_ignore_ascii_case(stringify!($v)) => Ok($e::$v),)+
-                    _                => Err({
-                                            let v = vec![
-                                                $(stringify!($v),)+
-                                            ];
-                                            format!("valid values:{}",
-                                                v.iter().fold(String::new(), |a, i| {
-                                                    a + &format!(" {}", i)[..]
-                                                }))
-                                        })
-                }
-            }
-        }
-
-        impl ::std::fmt::Display for $e {
-            fn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
-                match *self {
-                    $($e::$v => write!(f, stringify!($v)),)+
-                }
-            }
-        }
-
-        impl $e {
-            #[allow(dead_code)]
-            pub fn variants() -> Vec<&'static str> {
-                vec![
-                    $(stringify!($v),)+
-                ]
-            }
-        }
+    (enum $e:ident { $($v:ident),+ } ) => {
+        arg_enum!(@impls
+            (enum $e {
+                $($v),+
+            }) -> ($e, $($v),+)
+        );
     };
 }
 
-/// Allows you pull the version for an from your Cargo.toml as MAJOR.MINOR.PATCH_PKGVERSION_PRE
+/// Allows you pull the version for an from your Cargo.toml at compile time as
+/// MAJOR.MINOR.PATCH_PKGVERSION_PRE
 ///
 /// # Examples
+///
 /// ```no_run
 /// # #[macro_use]
 /// # extern crate clap;
 /// # use clap::App;
 /// # fn main() {
 ///     let m = App::new("app")
-///                 .version(&crate_version!()[..])
+///                 .version(crate_version!())
 ///                 .get_matches();
 /// # }
 /// ```
 #[macro_export]
 macro_rules! crate_version {
     () => {
-        env!("CARGO_PKG_VERSION").to_owned()
-    }
+        env!("CARGO_PKG_VERSION")
+    };
```

Comment:
Not per se, it's kind of like the trailing comma in structs and enums. It will be required if we add more rules to that macro.